### ISSUE TITLE
Annoying generic

### ISSUE THREAD

**Issue by szagoruyko** [torch/cutorch#100]

**Comment by hughperkins** [comment]
I think the dependencies do exist, just not being located.  I tried adding the `TH` path to include directories, and that's not sufficient, but I reckon something along those lines might somehow fix it.

```
INCLUDE_DIRECTORIES("${Torch_INSTALL_INCLUDE}/TH")
```

**Comment by hughperkins** [comment]
Rather than commenting out in make2cmake, one can comment out the following in `/usr/share/cmake-2.8/Modules/FindCUDA/run_nvcc.cmake` instead.  There might be consequences to this, but doing `touch` on a .cu file seems to trigger rebuild, when calling via `luarocks make`.

```
# Generate the dependency file
#cuda_execute_process(
#  "Generating dependency file: ${NVCC_generated_dependency_file}"
#  COMMAND "${CUDA_NVCC_EXECUTABLE}"
#  -M
#  ${CUDACC_DEFINE}
#  "${source_file}"
#  -o "${NVCC_generated_dependency_file}"
#  ${CCBIN}
#  ${nvcc_flags}
#  ${nvcc_host_compiler_flags}
#  ${depends_CUDA_NVCC_FLAGS}
#  -DNVCC
#  ${CUDA_NVCC_INCLUDE_ARGS}
#  )

#if(CUDA_result)
#  message(FATAL_ERROR "Error generating ${generated_file}")
#endif()

# Generate the cmake readable dependency file to a temp file.  Don't put the
# quotes just around the filenames for the input_file and output_file variables.
# CMake will pass the quotes through and not be able to find the file.
#cuda_execute_process(
#  "Generating temporary cmake readable file: ${cmake_dependency_file}.tmp"
#  COMMAND "${CMAKE_COMMAND}"
#  -D "input_file:FILEPATH=${NVCC_generated_dependency_file}"
#  -D "output_file:FILEPATH=${cmake_dependency_file}.tmp"
#  -P "${CUDA_make2cmake}"
#  )

#if(CUDA_result)
#  message(FATAL_ERROR "Error generating ${generated_file}")
#endif()

# Copy the file if it is different
#cuda_execute_process(
#  "Copy if different ${cmake_dependency_file}.tmp to ${cmake_dependency_file}"
#  COMMAND "${CMAKE_COMMAND}" -E copy_if_different "${cmake_dependency_file}.tmp" "${cmake_dependency_file}"
#  )

#if(CUDA_result)
#  message(FATAL_ERROR "Error generating ${generated_file}")
#endif()

# Delete the temporary file
#cuda_execute_process(
#  "Removing ${cmake_dependency_file}.tmp and ${NVCC_generated_dependency_file}"
#  COMMAND "${CMAKE_COMMAND}" -E remove "${cmake_dependency_file}.tmp" "${NVCC_generated_dependency_file}"
#  )

#if(CUDA_result)
#  message(FATAL_ERROR "Error generating ${generated_file}")
#endif()
```

The advantage that I see in modifying htis file, rather than make2cmake, is that it is more cuda-specific, and we can just include all these FindCUDA files inside torch, eg inside a `cmake` directory.

**Comment by netw1z** [comment]
was this ever solved? I get the same thing - install on Ubuntu on a raspberry pi.
The files are there:

    $locate THTensorCopy.h
 
    ~/torch/install/include/TH/generic/THTensorCopy.h
    ~/torch/pkg/torch/lib/TH/generic/THTensorCopy.h

But i get that stream of files not found. I edited cmake, but with no luck

**Comment by hughperkins** [comment]
Yes, I fixed it in https://github.com/torch/cutorch/pull/256 .  Please try, and confirm works ok, or else let me know any issues.

**Comment by hughperkins** [comment]
Note: how to try?  Like this:
```
git clone https://github.com/torch/cutorch.git
cd cutorch
git remote add hugh https://github.com/hughperkins/cutorch.git
git fetch hugh
git merge hugh/EXPERIMENTAL-improved-dependencies -m "merge"
luarocks make rocks/cutorch-scm-1.rockspec
luajit -l cutorch -e 'cutorch.test()'
```

**Comment by prashnani** [comment]
this works, thanks Hugh!

**Comment by lakehanne** [comment]
Thanks, @hughperkins .  This fixes me! :)

**Comment by hughperkins** [comment]
Cool.  @soumith what do you think are reasonable criteria to merge this?  Any particular tests you think it would be good for me to run?  Some kind of travis build perhaps?  I'm guessing for travis, the problem is not so much lack of gpu hardware, so much as, cuda toolkit installation license?

**Comment by hughperkins** [comment]
Not sure how well this will work on Mac (since I havent tested this on Mac), but you could try https://github.com/hughperkins/FindCUDA  It is based on cmake 3.5.0-rc1, ie the latest bleeding-edge version of cmake.  I've tested it using cmake 3.2.2, on ubuntu 15.10.

**Comment by hughperkins** [comment]
Update: the problem is now solvable very simply.  Simply do:

```
luarocks install FindCUDA
```

Thats it! From now on, your builds will be incremental, and will no longer produce streams of warning messages.

**Comment by lakehanne** [comment]
Thanks for your hard work !

**Comment by andreaskoepf** [comment]
@hughperkins what a relief, thank you so much! [FindCUDA](https://github.com/hughperkins/FindCUDA) was really an important 'patch' for every cutorch/cunn developer.

General questions: What is the state of this drop-in replacement for FindCuda.cmake? Is it a temporary solution until the official FindCuda gets updated? Does it have to be installed manually after a virgin torch installation via `luarocks install FindCuda`?

**Comment by hughperkins** [comment]
Thanks!

> Is it a temporary solution until the official FindCuda gets updated?

I'm not sure I'd label it 'temporary', since cmake wont be updated in LTS ubuntu until maybe 18.04?  It's possible it might be updated in time for 16.04, I havent really checked.  But I guess that 16.04 was taken from Debian sid (or equivalent) a few months ago, and therefore has the old version still?

So, I'd say more 'It is a solution until the official FindCuda gets updated, in ubuntu lts, which might not be until 18.04'

> Does it have to be installed manually after a virgin torch installation via luarocks install FindCuda?

It does.  Boris was pushing to have it added into 'distro'.  Personally, I tend to think that the subset of people who install distro who are going to hack on cutorch is pretty small, and maybe what we really want is something like a 'cutorch maintainers  faq', which might also reference FindCUDA install in passing?  But I dont have a strong opinion on this, either way is ok for me :-)

**Comment by hughperkins** [comment]
(Note that a 'cutorch maintainers faq' would sort of tie into this https://github.com/torch/cutorch/issues/70#issuecomment-189757152 perhaps? )

**Comment by hughperkins** [comment]
(Mind you, I have to say, that the installation is a lot more beautiful after installing FindCUDA . I've added FindCUDA to my own automated installations, so I dont have to watch page after page of spam scrolling past :-D

```
cd ~
wget https://github.com/hughperkins/FindCUDA/archive/v3.5-1.tar.gz -q -O ~/FindCUDA-v3.5-1.tar.gz
tar -xf FindCUDA-v3.5-1.tar.gz
cd ~/FindCUDA-3.5-1
luarocks make rocks/findcuda-scm-1.rockspec
rm ~/FindCUDA-v3.5-1.tar.gz
```

**Comment by hughperkins** [comment]
Seems like this is a solved problem now?

**Comment by elgalu** [comment]
I'm facing this issue, what's the solution this days? (3 years later :)
